Capability to specify per volume mount propagation mode

rhvgoyal opened the pull request:
Fixes #14630

This patch set allows one to specify mount propagation setting of a volume. Allowed values are "shared" or "slave" or "private".

For example.

docker run -ti -v /root/mnt-source:/root/mnt-dest:slave fedora bash

rhvgoyal commented:
CC @crosbymichael @cpuguy83 @LK4D4 @mrunalp @rhatdan

rhatdan commented:
If I specify two volumes one as shared and one as private

```+
+	if hasSharedVolume {
+		execdriver.SetRootPropagation(container, mount.SHARED)
+		return nil
+	}
+
+	if hasSlaveVolume {
+		execdriver.SetRootPropagation(container, mount.RSLAVE)
+		return nil
+	}
```
Won't propogation end up being RSLAVE?

rhvgoyal commented:
I first check if there are any shared volumes and set propagation to "shared" and return. I think you missed the "return" statement.

rhatdan commented:
Simpler would be to just check if a higher level is set and then don't change it.

if root prop = shared don't set to private, or slave,
If root prop == slave don't set to private

rhvgoyal commented:
@rhatdan I had thought about it. I think it is a good idea to implement it. Will do.

rhvgoyal commented:
- Took care of Dan's comments. Now modifying root propagation only if needed.
- Updated documentation of man pages.
- Exported Propagation status through docker-inspect Mounts.

rhvgoyal commented:
HOW TO TEST
============
- Make sure that parent mount of directory being bind mounted is `shared` (for shared volumes) or `shared/slave` (for slave volumes).

- Bind mount a directory as shared volume. 
$ docker run -ti /root/mnt-source:/root/mnt-dest:shared --cap-add SYS_ADMIN fedora bash

- Inside container make sure mount is shared.
$ findmnt -o TARGET,PROPAGATION /root/mnt-dest

- Do a mount under shared volume inside container
$ mkdir -p /root/mnt-dest/mnt1
$ cd /root/mnt-dest
$ mount --bind mnt1 mnt1

- Make sure this new mount propagates to host.
$ findmnt /root/mnt-source/mnt1

WHAT IS PARENT MOUNT
======================
- Run `df <directory>` and it will tell what's the parent mount. Looks for column `Mounted on`

HOW TO CHECK PROPAGATION PROPERTY OF  A MOUNT
================================================
$ findmnt -o TARGET,PROPAGATION <mount-point>

HOW TO CHANGE PROPAGATION PROPERTY OF A MOUNT
=================================================
$ mount --make-shared <mnt-point>
$ mount --make-slave <mnt-point>

crosbymichael commented on daemon/execdriver/native/create.go:
```diff
@@ -288,18 +289,42 @@ func (d *Driver) setupMounts(container *configs.Config, c *execdriver.Command) e
 
 	for _, m := range c.Mounts {
 		flags := syscall.MS_BIND | syscall.MS_REC
+		var propagationFlags []int
```
Why is this an array?

rhvgoyal replied:
In this case it does not have to be. Will change it to integer.

rhvgoyal committed:
Add capability to specify mount propagation per volume

Allow passing mount propagation option shared, slave, or private as volume
property.

For example.
docker run -ti -v /root/mnt-source:/root/mnt-dest:slave fedora bash

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal committed:
Check Propagation properties of source mount point

Whether a shared/slave volume propagation will work or not also depends on
where source directory is mounted on and what are the propagation properties
of that mount point. For example, for shared volume mount to work, source
mount point should be shared. For slave volume mount to work, source mount
point should be either shared/slave.

This patch determines the mount point on which directory is mounted and
checks for desired minimum propagation properties of that mount point. It
errors out of configuration does not seem right.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal commented:
@crosbymichael Fixed the array issue. Also I have added another patch which verifies that source mount point has appropriate propagation properties for volume mount to work, otherwise container start will fail. PTAL.

crosbymichael commented:
Looks like your additional change broke some stuff and there was a bad rebase

rhvgoyal commented:
@crosbymichael Found the bug I had introduced. PropagationFlags need to be nil if there is no propagation flag to be applied. With the second round of changes, it got initialized to 0 and libcontainer was trying to apply that and failing. Fixed it.

Also rebased on top of latest code. This volume support on windows moved quite a few things around.

crosbymichael commented:
Build error on windows binaries

rhvgoyal commented:
hmm..., Need to define volume.GetPropagation() for windows. Will do.

rhvgoyal commented:
Defined a function GetPropagation() for windows which returns nil string. Hopefully that fixes windows build. lets see.

jhowardmsft commented on volume/volume_windows.go:
```diff
@@ -179,3 +179,19 @@ func IsVolumeNameValid(name string) (bool, error) {
+	return rwModes[strings.ToLower(mode)]
+}
+
+// Windows does not support mount propagation. Just return empty string.
```
This should be godoc'd starting with the function name - something like "GetPropagation is a platform-specific function to ..... This is not supported on Windows".

rhvgoyal replied:
@jhowardmsft Will do. Thanks.

jhowardmsft commented:
Apart from the godoc nit, LGTM from a Windows perspective. @rhvgoyal Apologies for the rebase pain with the introduction of volume support on Windows.

rhvgoyal commented:
Took care of validate-lint issue. Added a comment starting with GetPropagation.

rhvgoyal commented:
@crosbymichael Did some more changes. PTAL.

- Added some unit and integration tests.
- Reorganized the code a bit. Propagation logic is specific to linux and not all unix. So moved it to volume/volume_propagation_linux.go. And made windows, freebsd and darwin use definition from volume/volume_propagation_unsupported.go.

rhvgoyal commented:
@crosbymichael Right now these are 3 patches. Do let me know if you want me to merge all this in single patch.

rhvgoyal commented:
My newly introduced integration test `TestRunVolumesMountedAsShared` is failing. It passes on my local machine (make test-integration-cli). so this has something to do with environment. I am not sure what though. Anybody has a clue what might be going on or what else can I try.

FAIL: docker_cli_run_test.go:3615: DockerSuite.TestRunVolumesMountedAsShared

docker_cli_run_test.go:3644:
    c.Fatalf("Failed to bind mount /volume-dest/mnt1: exit code %d", code)
... Error: Failed to bind mount /volume-dest/mnt1: exit code 255

rhvgoyal commented:
Ok, it was a permission issue. Running container now with "--privileged" and hopefully that resolves the issue and allows mount inside container.

rhvgoyal commented:
--privileged worked. In experimental build, both the tests passed.

PASS: docker_cli_run_test.go:3764: DockerSuite.TestRunVolumesMountedAsShared	0.400s
PASS: docker_cli_run_test.go:3800: DockerSuite.TestRunVolumesMountedAsSlave	0.367s

I think rest of the issues are not because of this patch set.

rhvgoyal commented:
Did some minor cleanup to tests. Did unmount operations for cleanup.

rhvgoyal commented:
@crosbymichael I think I am done with the changes now. PTAL.

albers commented on volume/volume_propagation_linux_test.go:
```diff
@@ -0,0 +1,49 @@
+		"/hostPath:/containerPath:ro,slave",
+		"/hostPath:/containerPath:ro,private",
+		"/hostPath:/containerPath:ro,z,shared",
+		"/hostPath:/containerPath:ro,Z,slave",
```
Do the mount options have to appear in the specified order?
Would `Z,slave,ro` and `slave,ro,Z` be legal syntax as well?
I'd like to see test cases that clarify this.
This is relevant for the shell completions.

rhvgoyal replied:
@albers No, they don't have to appear in specific order. Z,slave,ro is valid syntax too. I can add some more test cases for different order mount options.

albers replied:
@rhvgoyal Thanks, that would be great. Just some selected permutations would be great.

rhatdan replied:
Having multiple conflicting is bad.  But order does not matter.

Z,ro,z - bad
slave,Z,shared - Bad

rhvgoyal commented:
@albers added few more combinations to valid mount option combinations. PTAL.

rhvgoyal commented:
Windows build had failed as it did not have syscall.Umount(). Now I am using mount.Unmount() and hopefully that fixes it.

albers commented:
@rhvgoyal Thanks for adding the permutations to the test :smile:

rhvgoyal commented:
Rebased to latest upstream.

rhvgoyal commented:
Rebased to latest upstream.

rhvgoyal commented:
@crosbymichael ping.

peter-edge commented:
Now that Docker 1.9 is out, can we get this in ASAP? :)

rhvgoyal commented:
Rebased to resolve merge conflicts.

rhatdan commented:
@rhvgoyal Could you update the description to say that this pull request satisfies.

https://github.com/docker/docker/issues/14630

mrunalp commented:
@LK4D4 @crosbymichael PTAL :)

jhowardmsft commented on volume/volume.go:
```diff
@@ -51,7 +51,8 @@ type MountPoint struct {
 	Volume      Volume `json:"-"`
 
 	// Note Mode is not used on Windows
```
Can we also note that `Propagation` isn't used on Windows either.

rhvgoyal replied:
Will do.

ibuildthecloud commented:
Just throwing in my two cents that I need this PR soooooo much.  The amount of nsenter hacks I have to currently do is making even me feel bad :)

yasker commented:
Ack @ibuildthecloud . Badly need this. The mount point issue brought tons of trouble for volume plugin like Convoy, which has to do tons of nsenter to host in order to get volume works correctly.

wallnerryan commented:
+1!

crosbymichael commented:
exec driver stuff LGTM

ping @calavera for volumes

cpuguy83 commented:
See #17823 which this may end up making worse (especially if mounts are propagating from the host into the container's volume).

rhvgoyal commented:
@cpuguy83 I think we need to get to bottom of this issue. In general I think we should enforce all the mounts private (until and unless user overrides with propagation flag). My patch currently does not do that and I think I should add little more code to enforce a propagation default of private.

But this problem I need to understand. I have seen the cases of mounts leaking into containers and devices can't be deactivated as they are busy in other namespaces. But that's a problem specific to
devicemapper driver. For the case of filesystem based drivers (overlay etc), I think one should be able
to unmount conatiner root in daemon even if it leaked in some other mount namespace.

So yes, weared corner cases could arise and one will have to use this feature little carefully. It is no different then using mount namespace with mount propagation enabled. If not used properly, one
can run into various corner cases and kill the system.

vangie commented:
+1

rhvgoyal commented:
@crosbymichael @calavera @cpuguy83 pushed patches one more time. Now volume propagation will be `private` by default until and unless user overrides with `shared` or `slave`. PTAL.

rhvgoyal commented:
Pushed patches one more time. I realized that volume mounts by default are recursive binds. That means I also need to support recursive propagation properties like `rshared`, `rslave` and `rprivate`. Modified patches to support recursive propagation properties as well.

@crosbymichael @calavera PTAL. This feature has been outstanding for a very long time and we really want to see it available in 1.10.

SamYaple commented:
Indeed. We had it in 1.7-rc1 and it was ripped away from us! So close!

Though this implementation is far superior being that you can specify the propagation properties

thaJeztah commented:
sadly, looks like this needs a rebase

rhvgoyal commented:
Rebased.

rhvgoyal commented:
Any update on this PR. Can this be reviewed please.

thaJeztah commented:
ping @crosbymichael can you have another look at this?

cpuguy83 commented on daemon/container_unix.go:
```diff
@@ -1268,7 +1268,7 @@ func (container *Container) networkMounts() []execdriver.Mount {
 				Destination: "/etc/resolv.conf",
 				Writable:    writable,
-				Private:     true,
+				Propagation: volume.GetPropagation(""),
```
Instead of doing this, can we define a default propagation as a const?

rhvgoyal replied:
@cpuguy83 sure. Will define a constant global variable and use that.

rhvgoyal commented:
@cpuguy83 Pushed patches again. This time defined a const global variable for getting default propagation properties of a volume.

cpuguy83 commented:
1. When using a shared mount, creating a mount in a container... removing the container does not remove the mount... Should it?
2. I can't just create a host dir and mount stuff there from the container due to root mount issues -- or maybe this is because I'm using DinD to test?

Stinks we have to change the root mount propagation on the container to set the volume propagation.
Really hacky :(

cpuguy83 commented:
Playing around with it more, I think on point 1 we should unmount on container exit (or at least rm).

This is a really cool and powerful feature, but we should make sure to thoroughly document and help make sure users don't shoot themselves in the foot.

rhvgoyal commented:
@cpuguy83 For the first point, life time of mount on host should not be tied to container life time. One needs to run container again to unmount whenever they are done. So this is similar to running mount and unmount commands separately and only difference is that this time it is run from inside a container.

rhatdan commented:
Right we want to be able to do something like

# docker run --privileged -v /:/host nfsmount mount foobar:/mnt /host/mnt

Do local stuff on /mnt 

Later do

# docker run --privileged -v /:/host nfsmount umount foobar:/mnt /host/mnt

rhvgoyal commented:
@cpuguy83 W.r.t your second point, If I bind mount a directory from host inside container as 'shared' then I am able to mount something under that dir inside container and it becomes visible on host. I did  following.

$ ./docker run -ti -v /root/mnt-source:/root/mnt-dest:shared --cap-add SYS_ADMIN fedora bash
$ cd /root/mnt-dest
$ mkdir mnt1
$ mount --bind mnt1 mnt1

And now on host /root/mnt-source/mnt1 is a mount point.

- I am running docker daemon in host namespace and not in separate mount namespace. Otherwise
  shared mount will propagate to docker daemon mount namespace only and not host mount namespace.

If you could give me exact steps of how you are testing, I can give it a try here and see what's happening.

cpuguy83 commented:
@rhvgoyal `docker run -v /foo:/bar:shared busybox` yields an error that `/foo` is not shared (from the call to `ensureShared`).

`docker: Error response from daemon: Cannot start container b3fa8ee3df62fe3ef80f84d090ed990de1139bf4603bcf1ab72867dad2da702b: Path /foo is mounted on / but it is not a shared mount..`

cpuguy83 commented:
I'm assuming this is because I'm running docker in docker, where / is a private mount.

Works fine if I do something like
```
$ mount --make-shared /
$ docker run -v /foo:/bar:shared busybox
```

rhvgoyal commented:
@cpuguy83 Right. If parent mount of source is not shared, it can't be shared mount inside container. So I have put a patch to detect it error out if that's not the case. You can do two things to make it work.

- Convert /foo in a mount point and make it shared.
  $ mount --bind foo foo
  $ mount --make-shared foo

- Or convert / into a shared mount point. 
 mount --make-shared /
 
And in both the cases now source mount has "shared" properties and it should work.

rhvgoyal commented:
On fedora / is shared by default. So any directory which is bind mounted as shared volume works out of box without doing anything else.

rhvgoyal commented:
Rebased.

calavera commented:
LGTM

rhvgoyal commented:
@crosbymichael ping. Can this PR be merged now. 

You had given LGTM in the past and since then patches have changed only little.

cpuguy83 commented:
@rhvgoyal How would you feel about cleaning up mounts when the volume is removed, instead of as I suggested earlier (when the container was removed).

rhvgoyal commented:
@cpuguy83 So this applies to only docker managed internal volumes and plugin managed external volumes. right? Normal bind mounted volumes will not be subjected to this.

I can think of two options.

- Simple option is that volume removal fails as there are submounts under volume directory which might be in use by some programs.

- A more complicated option would be to do lazy unmounts of mounts under the volume directory and then try to delete the directory. I feel that we will run into various corner cases where if there are any programs running using those mounts or directories, these will be surprised.

So I feel that instead of this being default, it probably is better to implement some kind of additional options to volume removal where one could specify --force, and then volume cleanup logic goes on
to do forceful cleanup and possibly break any containers or programs on host using mounted directories.

In current form, I feel it is simplest to just leave it to user to cleanup volumes before these can be deleted if there are submounts underneath volume directories and volume removal will fail.

If cleanup of mounts is too big a concern, we could even think of limiting volume propagation only to 'bind mounted volumes' only and exclude docker created internal volumes and possibly volumes created using a plugin.

cpuguy83 commented:
As discussed on IRC, let's restrict this to binds only -- and explicitly _not_ named volumes (which look like a bind).

rhvgoyal commented:
@cpuguy83 Pushed patches again. Did following changes.

- One can specify volume propagation properties only for bind mounted volumes. For volumes backed by volume driver, one can not specify propagation properties.

- One can not specify propagation flag for internal volumes. That is the case even previous patches.

- Disallowed specifying  propagation properties for '--volumes-from'. One will inherit same volume properties as in data volume container.

- Added an optimization where all volume propgation flags are set to nil if rootfs propagation is RPRIVATE. That means anyway all volumes are going to be rpirvate so there is no bother runc to
apply these flags again per volume. Should save some computing overhead.

PTAL.

cpuguy83 commented on volume/volume_unix.go:
```diff
@@ -51,7 +43,8 @@ func ParseMountSpec(spec, volumeDriver string) (*MountPoint, error) {
 	mp := &MountPoint{
-		RW: true,
+		RW:          true,
+		Propagation: GetPropagation(""),
```
Should use default const instead of empty string.

rhvgoyal replied:
Will Change.

cpuguy83 commented:
@rhvgoyal Can you add integration tests for the binds/non-binds cases?

cpuguy83 commented:
Also ping @calavera if you have any extra comments here with the last change.

rhvgoyal commented:
@cpuguy83 Ok, will add integration test case for non-bind volume case.

calavera commented:
The additional guard looks great, thanks!

Still LGTM.

rhvgoyal commented:
@cpuguy83 Ok, pushed patches one more time.

- Got rid of GetPropagation("") call.
- Added bunch of unit tests for making sure propagation mode can not be passed with named volumes or internal volumes.
- Added one integration test for making sure propagation mode can not be specified with named volumes. I guess this might not be strictly needed due to unit test. Still having one integration test does not hurt in case code and unit tests change later.

cpuguy83 commented:
@rhvgoyal Perfect! Thanks!
Just waiting on tests and I'll move to docs review.
Code LGTM.

cpuguy83 commented:
LGTM, moving to docs review.

thaJeztah commented on man/docker-run.1.md:
```diff
@@ -508,6 +509,18 @@ content label. Shared volume labels allow all containers to read/write content.
+properties of a mount point use `mount` command. For example,
+`mount --make-shared /`.
+
 The `container-dir` must always be an absolute path such as `/src/docs`.
```
I think we should move this block (lines 524 - 535) more to the top (i.e. above `The **-v** option can be used one or...`. 

This section describes the basic usage of the `-v` flag, but is now after "advanced" usage (`:z/Z` and mount propagation)

rhvgoyal replied:
Ok, will do.

thaJeztah commented on man/docker-run.1.md:
```diff
@@ -508,6 +509,18 @@ content label. Shared volume labels allow all containers to read/write content.
 Only the current container can use a private volume.
 
+To control mount propagation property of volume one can use `:[r]shared`,
+`:[r]slave` or `:[r]private` propagation flag. Propagation property can
```
What's the difference between `shared` and `rshared` (also for the other options)?

rhvgoyal replied:
'rshared' is recursive while 'shared' is not. Same is for other options. 'r' means recursive. So if source dir has submounts underneath it, then these submounts will be visible in the container too and recursive will make sure all the submounts undergo through propagation flag operation instead of just top level mount.

thaJeztah commented on man/docker-run.1.md:
```diff
@@ -508,6 +509,18 @@ content label. Shared volume labels allow all containers to read/write content.
+To control mount propagation property of volume one can use `:[r]shared`,
+`:[r]slave` or `:[r]private` propagation flag. Propagation property can
+be specified only for bind mounted volumes and not for internal volumes or
+volume driver volumes. For this to work source mount point (mount point where
```
"mount point where source dir is mounted on"; so this is the `/var/lib/docker/volumes/......` path?

rhvgoyal replied:
No, it is not /var/lib/dokcer/volumes/. This directory container docker internal volumes and we don't allow passing propagation flags for docker internal volumes.

So it is the source mount of source-dir. So say you have a directory /root/foo which you want to bind mount inside container `-v /root/foo:/root/bar`. Then one can run `df` on `/root/foo` to figure out what's the source mount point. If /root/foo itself is a mount point, then source mount point is /root/foo. Otherwise most likely `/` is source mount point as /root/foo is mounted under `/`

thaJeztah commented on man/docker-run.1.md:
```diff
@@ -508,6 +509,18 @@ content label. Shared volume labels allow all containers to read/write content.
+source dir is mounted on) has to have right propagation properties. For shared
+volumes, source mount point has to be shared. And for slave volumes, source
+mount has to be either shared or slave. Use `df <source-dir>` to figure out
+the source mount and then use `findmnt -o TARGET,PROPAGATION <source-mount-dir>`
```
looks like the `findmnt` command is not available in boot2docker installs, so probably we need to mention an alternative, or instructions on how to install.

rhvgoyal replied:
One can also parse `/proc/self/mountinfo` and go through the mount entry of source mount. One needs to look at field 7 (optional fields) and see if there are any propagation properties specified for mount.

Following is from `man proc`.

`shared:X          mount is shared in peer group X
master:X          mount is slave to peer group X`

If nothing is specified, then mount point is private.

`findmnt` comes from util-linux package. I think it would be good if util-linux utilities are available in boot2docker. I have never used boot2dokcer. You will have to help me figure out how to make findmnt available on boot2docker. Otherwise we will have to resort to just explain looking at /proc/self/mountinfo which is not very easy.

thaJeztah replied:
Yup, boot2docker is probably an exception here, but many users run it, so we should think of that. 

ping @tianon any suggestions here?

thaJeztah commented on man/docker-run.1.md:
```diff
@@ -508,6 +509,18 @@ content label. Shared volume labels allow all containers to read/write content.
+the source mount and then use `findmnt -o TARGET,PROPAGATION <source-mount-dir>`
+ to figure out propagation properties of source mount. To change propagation
+properties of a mount point use `mount` command. For example,
+`mount --make-shared /`.
```
Should we use a different path than `/` here in the example? Will this mess up a user's system when copy/pasting this?

rhvgoyal replied:
A more safer example would be to just convert source directory into a mount point and then change its properties to shared. For example say `/root/foo` is source dir on the host. Then one can do following.

`mount --bind /root/foo /root/foo`
`mount --make-shared /root/foo`

And that will leave `/` untouched. I think I will mention both the methods in here.

Having said that we run with `/` being `shared` in fedora and things seem to be fine.

thaJeztah replied:
> Having said that we run with `/` being shared in `fedora` and things seem to be fine.

Yup, I understood that. But I wasn't sure if it would make a difference on other distros; people *will* copy/paste examples and we should prevent bringing them in trouble if not necessary.

thaJeztah commented:
I added some (probably n00b) questions inline; I realize this is an "advanced" topic, but still may need more explanation (and called out as being "advanced"?)

This documentation also needs to be added to the online documentation; https://github.com/docker/docker/blob/master/docs/reference/commandline/run.md and https://github.com/docker/docker/blob/master/docs/reference/run.md#volume-shared-filesystems

Perhaps we need to add a table at some point to specify all the available options to `-v` now that there are so many

thaJeztah commented:
Also, the option needs to be mentioned in https://github.com/docker/docker/blob/master/docs/reference/commandline/create.md and https://github.com/docker/docker/blob/master/man/docker-create.1.md (I think it's available there as well?)

rhvgoyal commented:
@thaJeztah Yes, these options are available for `create` too. I will add docs for create as well.

Will also add info into reference/commandline/run.md.

thaJeztah commented:
Thanks @rhvgoyal!

rhvgoyal commented:
@thaJeztah Existing `--volume` documentation in man/docker-create.1.md is just one line. So I am leaving it as it is and not adding documentation for mount propagation flags. People can refer to docker-run man page for details.

https://github.com/docker/docker/blob/master/man/docker-create.1.md

rhvgoyal commented:
@thaJeztah I think I have taken care of most of your comments. PTAL.

rhvgoyal commented:
@thaJeztah ping. Do latest patches look ok from documentation point of view?

thaJeztah commented:
@rhvgoyal Thanks for addressing my comments! I asked @moxiegirl if she has some bandwidth to have a look at them. I think they can have some TLC still (and she's far better at that than I am). I think she has enough here to work on now.

moxiegirl commented on daemon/execdriver/native/create.go:
```diff
@@ -319,31 +439,65 @@ func (d *Driver) setupMounts(container *configs.Config, c *execdriver.Command) e
+		// properties. If a volume is shared, then keep root propagtion
+		// shared. This should work for slave and private volumes too.
+		//
+		// For slave volumes, it can be either [r]shared/[r]slave.
```
Just out of curiosity, why do these options have such a complex format? Why not simply `shared`, `slave`, `private`? Why do we need the repeat of `[r]` --- if it is required by the underlying system why not handle it in the implementation?

thaJeztah replied:
options are shared, slave, private, rshared, rslave, rprivate (r for recursive); https://github.com/docker/docker/pull/17034#discussion_r47229306. We could consider mentioning them all separate.

rhvgoyal replied:
Right. There are total of six options. shared, slave and private only apply to top level of tree while recursive variants (rshared,  rslave, rprivate) apply to the whole fs tree underneath the directory.

moxiegirl replied:
Ah, that was so unclear.  I think we should break them out as individual options as per @thaJeztah .  If you are moving fast, it is just too hard to parse as is.  Thanks @rhvgoyal

moxiegirl commented on docs/reference/commandline/create.md:
```diff
@@ -80,11 +80,13 @@ Creates a new container.
       --ulimit=[]                   Ulimit options
       --uts=""                      UTS namespace to use
       -v, --volume=[]               Bind mount a volume with: [host-src:]container-dest[:<options>], where
```
@rhvgoyal Thank you for the contribution.  Here, the flag description should be short and sweet, a discussion of behavior level for later in the description.  Also, fill the square brackets with the actual format rather than leaving it to the flag statement...so:

```markdown
      -v, --volume=[host-src:]container-dest[:<options>]
                                    Bind mount a volume. The comma-delimited
                                    `options` are [rw|ro], [z|Z], or
                                    [[r]shared|[r]slave|[r]private]. The
                                    'host-src' is an absolute path or a name
                                    value.
```

rhvgoyal replied:
@moxiegirl Ok, will change it.

moxiegirl commented on docs/reference/commandline/run.md:
```diff
@@ -81,11 +81,13 @@ parent = "smn_cli"
       --uts=""                      UTS namespace to use
       -v, --volume=[]               Bind mount a volume with: [host-src:]container-dest[:<options>], where
-                                    options are comma delimited and selected from [rw|ro] and [z|Z].
+                                    options are comma delimited and selected from [rw|ro], [z|Z] and
```
Same here;

```markdown
      -v, --volume=[host-src:]container-dest[:<options>]
                                    Bind mount a volume. The comma-delimited
                                    `options` are [rw|ro], [z|Z], and
                                    [[r]shared|[r]slave|[r]private]. The
                                    'host-src' is an absolute path or a name
                                    value.
```

rhvgoyal replied:
@moxiegirl Will change.

moxiegirl commented on man/docker-run.1.md:
```diff
@@ -478,7 +478,8 @@ any options, the systems uses the following options:
 
 **-v**, **--volume**=[] Create a bind mount
```
`man/docker-create.1.md` is missing

rhvgoyal replied:
@moxiegirl man/docker-create.1.md has a very basic description of volume. It does not even specify options ro/rw or z/Z. So it does not look like it has kept pace with docker-run.1.md.That's why I left it untouched.

I will change the --volume description as you have suggested.

moxiegirl commented on man/docker-run.1.md:
```diff
@@ -478,7 +478,8 @@ any options, the systems uses the following options:
 
 **-v**, **--volume**=[] Create a bind mount
    (format: `[host-dir:]container-dir[:<suffix options>]`, where suffix options
```
```markdown

[**-v**|**--volume**[=*[[HOST-DIR:]CONTAINER-DIR[:OPTIONS]]*]]

  Create a bind mount. If you specify, ` -v /HOST-DIR:/CONTAINER-DIR`, Docker bind mounts `/HOST-DIR` in the host to `/CONTAINER-DIR` in the Docker container. If 'HOST-DIR' is omitted,  Docker automatically creates the new volume on the host.  The `OPTIONS` are a comma delimited list and can be:

  * [rw|ro]
  * [z|Z]
  * [`[r]shared`|`[r]slave`|`[r]private`]


The `CONTAINER-DIR` must be an absolute path such as `/src/docs`. The `HOST-DIR`
can be an absolute path or a `name` value. A `name` value must start with an
alphanumeric character, followed by `a-z0-9`, `_` (underscore), `.` (period) or `-` (hyphen). An absolute path starts with a `/` (forward slash).

If you supply a `HOST-DIR` that is an absolute path,  Docker bind-mounts to the
path you specify. If you supply a `name`, Docker creates a named volume by that
`name`. For example, you can specify either `/foo` or `foo` for a `HOST-DIR`
value. If you supply the `/foo` value, Docker creates a bind-mount. If you
supply the `foo` specification, Docker creates a named volume.

You can specify multiple  **-v** options to mount one or more mounts to a
container. To use these same mounts in other containers, specify the
**--volumes-from** option also.
```

rhvgoyal replied:
ok, I will change the documentation as per your suggestion.

moxiegirl commented:
@rhvgoyal Unfortunately, I don't agree with the no updates to `docker-create.1.md`.  I realize copy paste is a pain and so is the duplication. But most users expect man pages to be complete descriptions.  We should only point them to an external page if that page has a related command.

Also, this is a pretty sophisticated topic and an important one, should add the overview information here:

https://docs.docker.com/engine/userguide/dockervolumes/

rhvgoyal commented:
@moxiegirl 

Also look at some of the steps I provided in this thread about how to test it.
https://github.com/docker/docker/pull/17034#issuecomment-148785874

Here are the use cases.

By default bind mounted volumes are `private`. That means any mount operation done on host after the container has launched does not become visible inside container. And similarly any mount operation done inside container does not become visible on host. And these additional volume options allow one to change default behavior based on user's need.

For example, one use case is that we want to mount say ceph/gluster volumes on host so that these volumes can be used by other containers. Utilities to mount these volumes might not be available on host and can be shipped as part of a container. So actual mounting of the volume will take place from inside the container but the mount needs to be visible on host so that it can be further bind mounted inside other containers. As of today it is not possible and any mounts done inside containers are not visible on host. IOW, mounts inside container do not propagate to parent mount namespace.

With these patches, one can bind mount a directory on host as `shared` and that means mounts done under that volume inside container will show up on host as well and vice-a-versa. For example, say you mount directory on host /root/mnt-source inside a container as follows.

`docker run -ti /root/mnt-source:/root/mnt-dest:shared --cap-add SYS_ADMIN fedora bash`

Now if container performs another mount operation under /root/mnt-dest, it should propagate to host and new mount point should become visible under /root/mnt-source/. For example, create a directory and convert it in a mount point. Do following inside container.

`mkdir /root/mnt-dest/mnt1`
`mount --bind mnt1 mnt1`

Now `mnt1` is a mount point and you will see this mount point on host as `/root/mnt-source/mnt1`. In practice this will be an actual mount mounting a block device so that `/root/mnt-source/mnt1` can be used as source dir for bind mounted volume for another container.

`docker run -ti /root/mnt-source/mnt1:/data fedora bash`

`slave` volumes allow only one directional propagation. That is if host mounts something it becomes visible under container but if container mounts something it does not become visible on host. This mode is being considered useful for things like `autofs`. Idea is that as soon as container accesses some directory, autofs on host will mount it and that mount point will become visible inside container.

`private` mode is current default and that is no mount propagation in any direction. Mounts inside container don't become visible on host and vice-a-versa.

For more details about mount propagation, you can also refer to following documentation in kernel. After a while it becomes little dense though.

`https://github.com/torvalds/linux/blob/master/Documentation/filesystems/sharedsubtree.txt`

rhvgoyal committed:
Add some unit and integration tests

Add a unit test and couple of integration tests for volume propagation.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal commented:
@moxiegirl I think I have taken care of your review comments so far. PTAL.

moxiegirl commented:
@rhvgoyal you rock, that's wonderful material.  It definitely belongs in the volumes material.  Let me do an update and turn it around to you with ways to ref from the command and the man pages.

moxiegirl commented:
@rhvgoyal I'm really liking the use case you provided. Really will help to add it to the docs. Looking at our volumes docs there is a lot of cruft in there that needs to be sorted. Rather than hold this PR up, I'm happy to carry the docs if you are willing to come back in with a review -- because I will have incorporated your excellent use case and I'll need your eyes on it to make sure I haven't borked it up in placing it.  

So, docs LGTM with the proviso that you'll do eyes on the new carry pr.

rhvgoyal commented:
@moxiegirl Sure, I will review the new PR for documenting this feature better. That's not a problem at all.

cpuguy83 commented:
Awesome, merging!

rhvgoyal commented:
Thanks @cpuguy83

peter-edge commented:
![image](https://cloud.githubusercontent.com/assets/4228796/11818297/e91c460c-a359-11e5-8c3a-45177ff70010.png)

thaJeztah commented:
Cool, thanks @rhvgoyal, and thanks @cpuguy83 for getting it moving :fireworks:

mrunalp commented:
Yay 😃

Sent from my iPhone

> On Dec 15, 2015, at 9:26 AM, Vivek Goyal <notifications@github.com> wrote:
> 
> Thanks @cpuguy83
> 
> —
> Reply to this email directly or view it on GitHub.
>

bnoon commented:
Great!!

rhatdan commented:
Finally....   This is a great feature.

SamYaple commented:
Awesome!

griwes commented:
Oh look, this is exactly what I need for a thing I'm doing at this moment. When can a release including this functionality be expected to appear?

thaJeztah commented:
@griwes this will be in the 1.10 release, probably in February. If you want to give it a test, you can either try the builds from master (https://master.dockerproject.org) or the experimental builds https://experimental.docker.com.

lox commented:
Is it assumed knowledge that `/root/mnt-source` in your example needs to be mounted as "shared" on the host before this will work?

```
$ sudo mount --bind /root/mnt-source /root/mnt-source
$ sudo mount --make-shared /root/mnt-source
```

Without this I got an error indicating that `/root/mnt-source` wasn't shared (under Ubuntu 14.04 with the latest experimental docker).

rhatdan commented:
Yes this is expected.  Note that RHEL7, Fedora, And Centos come this way by default.

2opremio commented on man/docker-run.1.md:
```diff
@@ -508,18 +518,36 @@ content label. Shared volume labels allow all containers to read/write content.
+
+To change propagation properties of a mount point use `mount` command. For
+example, if one wants to bind mount source directory `/foo` one can do
+`mount --bind /foo /foo` and `mount --make-private --make-shared /foo`. This
```
shouldn't this be `mount --make-shared` (i.e. without `--make-private`)?

rhvgoyal replied:
Usage of `--make-private` decouples the mount from any other sharing already going on. So while this is not strictly necessary, this might be little less confusing in some of the configurations when propagation actually happens.

kiwenlau commented:
@lox 

How to set following permanently? I have to re mount it after reboot the os

```
sudo mount --bind /root/mnt-source /root/mnt-source
sudo mount --make-shared /root/mnt-source
```

Thx

rhatdan commented:
Make a unit file or service script to do this.

SamYaple commented:
@kiwenlau on 14.10+ with systemd you can set the MountFlags=shared and you wont need to set --make-shared each time. You can also set that in /etc/fstab with 14.10+.

on 14.04 you are limited to some type of init script for this (can also be a udev trigger depending on if that makes the situation better)

kiwenlau commented:
@rhatdan @SamYaple I use Ubuntu 14.04.
I can set mount in /etc/fstab:
`/root/mnt-sourc         /root/mnt-sourc             none    bind            0       0`

However, I can not set shared mount for it in fstab file. I have to manually set it when I reboot ubuntu
`sudo mount --make-shared /root/mnt-source`

It's OK for me, and I will try some init script to do it. Thanks.

SamYaple commented:
@kiwenlau Yes i believe I stated that this only works on 14.10+ when set in /etc/fstab or with the mount binary.

If you pullin the mount binary from 14.10+ it would work, but may break other things
